packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts

```diff
@@ -162,24 +218,33 @@ class GrpcGatewayListener extends VirtualGatewayListener {
    */
   protected protocol: Protocol = Protocol.GRPC;
 
-  constructor(options: HttpGatewayListenerOptions = {}) {
+  /**
+   * Represents the listener certificate
+   *
+   * @default - none
+   */
+  protected readonly tlsCertificate?: TlsCertificate;
+
+  constructor(props: HttpGatewayListenerOptions = {}) {
     super();
-    this.port = options.port ? options.port : 8080;
-    this.healthCheck = options.healthCheck;
+    this.port = props.port ? props.port : 8080;
+    this.healthCheck = props.healthCheck;
+    this.tlsCertificate = props.tlsCertificate;
   }
 
   /**
    * Called when the GatewayListener type is initialized. Can be used to enforce
    * mutual exclusivity
    */
-  public bind(_scope: cdk.Construct): VirtualGatewayListenerConfig {
+  public bind(scope: cdk.Construct): VirtualGatewayListenerConfig {
     return {
       listener: {
         portMapping: {
           port: this.port,
           protocol: Protocol.GRPC,
         },
         healthCheck: this.healthCheck ? renderHealthCheck(this.healthCheck, this.protocol, this.port): undefined,
+        tls: this.tlsCertificate ? renderTls(scope, this.tlsCertificate) : undefined,
```

Comment:
I really hate all of this duplication in this file 😕 (I understand it was all here before your changes). Can we change it while you're in the area?

Let's have a single class, `VirtualGatewayListenerImpl`, that takes the individual properties in the constructor, and then let's return that class from `grpc()`,  `http()` and `http2()` in `VirtualGatewayListener`.

Look at how this is implemented in [`VirtualNodeListener` class](https://github.com/aws/aws-cdk/blob/56886d0ed38736cb87cf0d26aede5e89432157f2/packages/%40aws-cdk/aws-appmesh/lib/virtual-node-listener.ts#L124-L124) - model your solution after that file, as it nicely removed all duplication.